src/PluginClientBuilder.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Http\Client\Common;
+
+use Http\Client\HttpClient;
+use Http\Client\HttpAsyncClient;
+use Psr\Http\Client\ClientInterface;
+
+/**
+ * Build an instance of a PluginClient with a dynamic list of plugins.
+ *
+ * This builder is immutable, each "build" methods will instead create a new
+ * instance of itself with or without the affected plugin, destroying a
+ * potentially http client built in the new instance.
+ *
+ * @author Baptiste Clavié <clavie.b@gmail.com>
+ */
+final class PluginClientBuilder
+{
+    /** @var Plugin[] */
+    private $plugins = [];
+
+    /** @var ClientInterface | HttpClient | HttpAsyncClient */
+    private $client;
+
+    /** @var ?PluginClient */
+    private $pluginClient;
+
+    /**
+     * @param ClientInterface | HttpClient | HttpAsyncClient $client Client to nest into the plugin client
+     */
+    public function __construct($client)
+    {
+        if (!$client instanceof ClientInterface && !$client instanceof HttpClient && !$client instanceof HttpAsyncClient) {
+            throw new \RuntimeException('You must provide a valid http client');
+        }
+
+        $this->client = $client;
+    }
+
+    public function addPlugin(string $name, Plugin $plugin): self
+    {
+        $builder = new self($this->client);
+
+        $builder->pluginClient = null;
+        $builder->plugins = $this->plugins;
+        $builder->plugins[$name] = $plugin;
+
+        return $builder;
+    }
+
+    public function removePlugin(string $name): self
+    {
+        $builder = new self($this->client);
+
+        $builder->pluginClient = null;
+        $builder->plugins = $this->plugins;
+        unset($builder->plugins[$name]);
+
+        return $builder;
+    }
+
+    public function createClient(array $options = []): PluginClient
```

Comment:
i would prefer a `setOption(string $name, $value)` (and `removeOption(string $name)`) method instead. we right now don't have many options in the PluginClient, but it would be consistent imho.

Reply:
I'd rather have to pass the client and the options on the `createClient` for now, especially as we're building a plugin client which has few (one) options.

Adding those setters can be useful indeed, but I don't think it is needed right now, right ?

Reply:
i know there is only one option at the moment... but this way is not future proof. if we start adding options, having setOption / removeOption *and* the $options for createClient will be confusing.